src/net/server/handlers/message_stream_layer/service.rs

```diff
@@ -0,0 +1,134 @@
+    }
+}
+
+pub struct ReservedPermit<T>(PollSender<T>);
+impl<T: Send + 'static> ReservedPermit<T> {
+    pub fn send(&mut self, item: T) -> Result<(), MpcServerError> {
+        Ok(self.0.send_item(item)?)
+    }
+}
+
+/// `poll_ready` will first grab a permit via the `PollSender`. Once ready, the sender is injected
+/// into the Request as an Extension. This can only be done once
+#[derive(Debug)]
+pub struct MessageStream<S, T> {
+    inner: S,
+    sender: SenderStatus<T>,
+}
+
+impl<S, T: Send + 'static> MessageStream<S, T> {
+    pub fn new(inner: S, sender: mpsc::Sender<T>) -> Self {
+        Self {
+            inner,
+            sender: SenderStatus::NotReady(PollSender::new(sender)),
+        }
+    }
+}
+
+/// Within `poll_ready`, first grabs a permit for the [`PollSender`]. Then, pass the [`PollSender`]
+/// into the request as an [`Extension`] on the `call`.
+///
+/// Leaves the [`Service`]'s `Error` type unspecified, so that it can be [`Infallible`] within the
+/// context of Axum. That additionally means never returning an error, instead creating a response
+/// with an error [`StatusCode`]
+impl<S, ReqBody, ResBody, T> Service<Request<ReqBody>> for MessageStream<S, T>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>>,
+    ResBody: Default,
+    T: Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = ResponseFuture<S::Future>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        match &mut self.sender {
+            SenderStatus::NotReady(sender) => {
+                if ready!(sender.poll_reserve(cx)).is_ok() {
+                    // swap `NotReady` for `Ready`
+                    if let SenderStatus::NotReady(sender) =
+                        std::mem::replace(&mut self.sender, SenderStatus::Spent)
+                    {
+                        self.sender = SenderStatus::Ready(sender);
+                        self.inner.poll_ready(cx)
+                    } else {
+                        panic!("SenderStatus was not NotReady")
+                    }
+                } else {
+                    // if `poll_reserve` returns error, forward error to response
+                    self.sender = SenderStatus::Err;
+                    Poll::Ready(Ok(()))
+                }
+            }
+            SenderStatus::Ready(_) => self.inner.poll_ready(cx),
+            SenderStatus::Err | SenderStatus::Spent => Poll::Ready(Ok(())),
+        }
+    }
+
+    fn call(&mut self, mut req: Request<ReqBody>) -> Self::Future {
+        match &self.sender {
+            SenderStatus::NotReady(_) => panic!("poll_ready not called"),
+            SenderStatus::Ready(_) => {
+                if let SenderStatus::Ready(sender) =
+                    std::mem::replace(&mut self.sender, SenderStatus::Spent)
+                {
+                    // must insert value wrapped in `Some` due to `Extension`'s `FromRequest` impl
```

Comment:
You don't need that because you can just borrow the body (`body_mut`) inside the handler, convert it to owned vec of envelopes and only then borrow the extension mutably and send.

Reply:
oh good. much better...